File: dat-ruby/lib/saro/dat/dat_manager.rb
# frozen_string_literal: true

require 'set'
require_relative 'dat_certificate'
require_relative 'dat'
require_relative 'signature'
require_relative 'util'

module Saro
  module Dat
    class DatManager
      # Immutable snapshot of the manager state.
      # Readers access it lock-free via a single ivar read (atomic reference swap);
      # writers rebuild a new frozen snapshot under @write_lock.
      State = Struct.new(:issuer, :certificates, :by_cid)

      EMPTY_STATE = State.new(nil, [].freeze, {}.freeze).freeze
      private_constant :EMPTY_STATE

      def initialize
        @state = EMPTY_STATE
        @write_lock = Mutex.new
      end

      def import_certificates(input_certs, clear: false)
        renew_count = 0
        @write_lock.synchronize do
          certificates = clear ? [] : @state.certificates.dup

          before_cids = Set.new(certificates.map(&:cid))
          seen_cids = Set.new

          input_certs.each do |cert|
            raise ArgumentError, "Duplicate CID: #{cert.cid}" if seen_cids.include?(cert.cid)
            seen_cids.add(cert.cid)
            next if cert.expired
            next if before_cids.include?(cert.cid)

            certificates << cert
            renew_count += 1
          end

          certificates.sort_by!(&:dat_issuance_end_seconds)

          # Find latest issuable certificate as issuer
          issuer = certificates.reverse_each.find(&:issuable)

          by_cid = {}
          certificates.each { |cert| by_cid[cert.cid] = cert }

          @state = State.new(issuer, certificates.freeze, by_cid.freeze).freeze
        end
        renew_count
      end

      def imports(format_str, clear: false)
        certs = []
        format_str.strip.split("\n").each do |line|
          line = line.strip
          next if line.empty?
          certs << Saro::Dat::DatCertificate.imports(line)
        end
        import_certificates(certs, clear: clear)
      end

      def exports(verify_only = false)
        @state.certificates.map { |cert| cert.exports(verify_only) }.join("\n")
      end

      def issue(plain, secure)
        issuer = @state.issuer
        raise RuntimeError, "Invalid DAT: Signing Key Does Not Exist" unless issuer

        self.class._issue(issuer, plain, secure)
      end

      def parse(dat_input)
        dat = Saro::Dat::Dat.from_value(dat_input)
        raise ArgumentError, "Invalid DAT: Format" unless dat.format

        certificate = @state.by_cid[dat.cid]

        raise ArgumentError, "Invalid DAT: CID(Certificate ID) Not Found" unless certificate

        self.class._parse(certificate, dat)
      end

      private

      def self._issue(cert, plain, secure)
        now = Time.now.to_i
        expire = now + cert.dat_ttl_seconds
        cid_hex = cert.cid.to_s(16)

        plain_b64 = Saro::Dat::Util.encode_base64_url_str(plain)

        encrypted_secure = cert.crypto_key.encrypt(secure)
        secure_b64 = Saro::Dat::Util.encode_base64_url_str(encrypted_secure)

        body = "#{expire}.#{cid_hex}.#{plain_b64}.#{secure_b64}"
        signature = Saro::Dat::Util.encode_base64_url_str(cert.signature_key.sign(body))

        "#{body}.#{signature}"
      end

      def self._parse(cert, dat_input)
        dat = Saro::Dat::Dat.from_value(dat_input)
        raise RuntimeError, "Invalid DAT: Format" unless dat.format
        raise RuntimeError, "Invalid DAT: Expired" if dat.expired?

        unless cert.signature_key.verify(dat.body_string, dat.signature)
          raise RuntimeError, "Invalid DAT: Signature"
        end

        decrypted_secure = cert.crypto_key.decrypt(dat.secure)
        Saro::Dat::DatPayload.new(dat.plain, decrypted_secure)
      end
    end
  end
end

File: dat-pypi/src/saro_dat/dat_manager.py
from __future__ import annotations
import threading
import time
from typing import List, Optional, Union

from . import DatCertificate, Dat, DatPayload
from .util import encode_base64_url_str


class DatManager:
    def __init__(self):
        self._issuer: Optional[DatCertificate] = None
        self._certificates: tuple = ()
        self._certificates_by_cid: dict = {}
        self._write_lock = threading.Lock()


    def import_certificates(self, input_certs: List[DatCertificate], clear: bool = False) -> int:
        with self._write_lock:
            certificates = [] if clear else list(self._certificates)
            renew_count = 0

            before_cids = set(map(lambda x: x.cid, certificates))
            seen_cids = set()

            for cert in input_certs:
                if cert.cid in seen_cids:
                    raise ValueError(f"Duplicate CID: {cert.cid}")
                seen_cids.add(cert.cid)
                if cert.expired():
                    continue
                if cert.cid in before_cids:
                    continue
                certificates.append(cert)
                renew_count += 1

            certificates.sort(key=lambda x: x.dat_issuance_end_seconds)
            issuer = next((c for c in reversed(certificates) if c.issuable()), None)

            self._issuer = issuer
            self._certificates = tuple(certificates)
            self._certificates_by_cid = {c.cid: c for c in certificates}

            return renew_count


    def imports(self, format_str: str, clear: bool = False) -> int:
        certs = []
        for line in format_str.strip().split('\n'):
            if line.strip():
                certs.append(DatCertificate.imports(line.strip()))
        return self.import_certificates(certs, clear)

    def exports(self, verify_only: bool = False) -> str:
        lines = []

        for cert in self._certificates:
            if not verify_only or cert.support_verify_only():
                lines.append(cert.exports(verify_only))

        return '\n'.join(lines)

    def _find_unsafe(self, cid: int) -> Optional[DatCertificate]:
        return self._certificates_by_cid.get(cid)

    def issue(self, plain: Union[bytes, str, None], secure: Union[bytes, str, None]) -> str:
        issuer = self._issuer
        if issuer:
            return self._issue(issuer, plain, secure)
        raise RuntimeError("Invalid DAT: Signing Key Does Not Exist")

    def parse(self, dat_input: Union[Dat, str, None]) -> DatPayload:
        dat = Dat.from_value(dat_input)
        if not dat._format:
            raise ValueError("Invalid DAT: Format")

        certificate = self._certificates_by_cid.get(dat._cid)
        if certificate is not None:
            return self._parse(certificate, dat)
        raise ValueError("Invalid DAT: CID(Certificate ID) Not Found")

    @staticmethod
    def _issue(cert: DatCertificate, plain: Union[bytes, str, None], secure: Union[bytes, str, None]) -> str:
        now = int(time.time())
        expire = now + cert.dat_ttl_seconds
        cid_hex = hex(cert.cid)[2:]

        # Plain 데이터 처리 (문자열인 경우 utf-8 바이트로)
        plain_bytes = plain.encode() if isinstance(plain, str) else (plain or b'')
        plain_b64 = encode_base64_url_str(plain_bytes)

        # Secure 데이터 암호화
        encrypted_secure = cert._crypto_key.encrypt(secure)
        secure_b64 = encode_base64_url_str(encrypted_secure)

        body = f"{expire}.{cid_hex}.{plain_b64}.{secure_b64}"
        signature = encode_base64_url_str(cert._signature_key.sign(body))

        return f"{body}.{signature}"

    @staticmethod
    def _parse(cert: DatCertificate, dat_input: Union[Dat, str, None]) -> DatPayload:
        dat = Dat.from_value(dat_input)
        if not dat._format:
            raise RuntimeError("Invalid DAT: Format")
        if dat.expired():
            raise RuntimeError("Invalid DAT: Expired")

        # 서명 검증
        if not cert._signature_key.verify(dat.body_string(), dat._signature):
            raise RuntimeError("Invalid DAT: Signature")

        # 데이터 복호화
        decrypted_secure = cert._crypto_key.decrypt(dat._secure)
        return DatPayload(dat._plain, decrypted_secure)
